### PR TITLE
Feature/improved takeoff

### PR DESCRIPTION
**Pull request by Thomas-McKanna**
This pull request adds the use of simple_takeoff to TakeoffTask. Because simple_takeoff is a hassle to get working in the simulator (it requires opening up MissionPlanner and setting EKF origin each time), I have added a TakeoffTaskSim class which is chosen when the code is running in Gazebo. All that needs to be done is set is_simulator=True when constructing the DroneController object.

### CODE REVIEW

File: flight/AIs/user_command.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 
 def main():
     # Make the controller object
-    controller = DroneController(c.Drones.LEONARDO_SIM)
+    controller = DroneController(c.Drones.LEONARDO_SIM, is_simulation=True)
```
Comment (christopher-o-toole) [inline flight/AIs/user_command.py]: hardcoded reeeee. Pass in a command-line argument here.

File: flight/drone/drone_controller.py [hunk 2]
```diff
@@ -32,18 +34,23 @@ class DroneController(object):
         A PriorityQueue holding tasks to be performed.
     _safety_event : Event
         Set when an unsafe condition is observed.
+    _is_simulation : bool
+        Set to true when the code is intended for the simulator.
     """
 
-    def __init__(self, drone):
+    def __init__(self, drone, is_simulation=False):
```
Comment (christopher-o-toole) [inline flight/drone/drone_controller.py]: Same applies here, use a command-line argument for easy configuration

File: flight/drone/drone_controller.py [hunk 3]
```diff
@@ -32,18 +34,23 @@ class DroneController(object):
...
         """Construct a drone controller.
 
         Parameters
         ----------
-        drone : c.Drone.{DRONE_NAME}"""
+        drone : c.Drone.{DRONE_NAME}
```
Comment (christopher-o-toole) [inline flight/drone/drone_controller.py:54]: Undocumented parameter is_simulation (also make sure to document the default value for this parameter)

File: flight/drone/drone_controller.py [hunk 4]
```diff
@@ -32,18 +34,23 @@ class DroneController(object):
...
+        """
         self._task_queue = PriorityQueue()
         self._current_task = None
         self._safety_event = Event()
 
+        self._is_simulation = is_simulation
+
         # Initialize the logger
         self._logger = logging.getLogger(__name__)
         coloredlogs.install(level=logging.INFO)
```
Comment (christopher-o-toole) [inline flight/drone/drone_controller.py]: Logging-level should either be a file constant or be configurable through an argument passed to this function (preferably the latter)

File: flight/tasks/takeoff_task.py [hunk 5]
```diff
@@ -21,16 +32,19 @@ def __init__(self, drone, altitude):
         """
         super(TakeoffTask, self).__init__(drone)
         self._target_alt = altitude
+        self._count = 10 / c.DELAY_INTERVAL
+        self._began_takeoff = False
+
     def perform(self):
```
Comment (christopher-o-toole) [inline flight/tasks/takeoff_task.py:49]: Missing function docstring.

File: flight/tasks/takeoff_task_sim.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            self._drone.arm()
+
+        current_altitude = self._drone.rangefinder.distance
+
+        if current_altitude >= self._target_alt * f.PERCENT_TARGET_ALTITUDE:
+            return True
+
+        thrust = f.DEFAULT_TAKEOFF_THRUST
+
+        self._drone.set_attitude(0, 0, 0, thrust)
```
Comment (christopher-o-toole) [inline flight/tasks/takeoff_task_sim.py]: these zero values should be configurable parameters which default to zero.

File: flight/tasks/takeoff_task_sim.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        ----------
+        drone : dronekit.Vehicle
+            The drone being controlled
+        altitude : float
+            How many meters off the ground to take off to.
+        """
+        super(TakeoffTaskSim, self).__init__(drone)
+        self._target_alt = altitude
+
+    def perform(self):
```
Comment (christopher-o-toole) [inline flight/tasks/takeoff_task_sim.py:35]: Missing function docstring.

File: flight/tasks/takeoff_task_sim.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
+from task_base import TaskBase
```
Comment (christopher-o-toole) [inline flight/tasks/takeoff_task_sim.py:6]: All new files should have docstrings from now on.

**Review by christopher-o-toole — CHANGES_REQUESTED** [review]
Decent, but remember to start using docstrings for each new file from now on and don't forget to add docstrings for class methods/functions. Also, is_simulator should be a command-line argument passed when the application is ran.

File: flight/AIs/user_command.py (hunk 1 above)
Comment (ClayMav) [reply flight/AIs/user_command.py]:
```suggestion
    controller = DroneController(is_simulation=True)
```

File: flight/drone/drone_controller.py (hunk 2 above)
Comment (ClayMav) [reply flight/drone/drone_controller.py]:
```suggestion
    def __init__(self, is_simulation=False):
```

File: flight/AIs/user_command.py (hunk 1 above)
Comment (ClayMav) [reply flight/AIs/user_command.py]: Lets set the SIM_DRONE and DRONE as constants and just inherit which drone to use based on this flag

File: flight/drone/drone_controller.py (hunk 4 above)
Comment (Thomas-McKanna) [reply flight/drone/drone_controller.py]: Added

File: flight/tasks/takeoff_task_sim.py (hunk 6 above)
Comment (Thomas-McKanna) [reply flight/tasks/takeoff_task_sim.py]: Added

File: flight/AIs/user_command.py [hunk 9]
```diff
@@ -9,8 +9,17 @@
 PROMPT_FOR_COMMAND = '> '
 
 def main():
+    if len(sys.argv) == 2:
```
Comment (christopher-o-toole) [inline flight/AIs/user_command.py]: Use https://docs.python.org/2/library/argparse.html

File: flight/tasks/takeoff_task.py [hunk 10]
```diff
@@ -21,16 +32,19 @@ def __init__(self, drone, altitude):
         """
         super(TakeoffTask, self).__init__(drone)
         self._target_alt = altitude
+        self._count = 10 / c.DELAY_INTERVAL
```
Comment (christopher-o-toole) [inline flight/tasks/takeoff_task.py]: hardcoded 10 gives me vietnam flashbacks.

**Review by christopher-o-toole — CHANGES_REQUESTED** [review]
A few things must still be changed.

**Review by ClayMav — CHANGES_REQUESTED** [review]

File: flight/AIs/user_command.py (hunk 9 above)
Comment (ClayMav) [reply flight/AIs/user_command.py]: unnecessary.

File: flight/AIs/user_command.py [hunk 11]
```diff
@@ -22,6 +26,16 @@ def main():
...
     controller.run()
 
+def create_parser():
+    """Returns a configured argument parser."""
+    parser = argparse.ArgumentParser(description='Test flight commands.')
+    parser.add_argument('--sim',
+                        dest='sim',
+                        action='store_true',
+                        default=False,
+                        help='run simulator compatable flight code')
```
Comment (christopher-o-toole) [inline flight/AIs/user_command.py]:
```suggestion
                        help='run simulator compatible flight code')
```

**Comment by Thomas-McKanna** [comment]
@ClayMav Is this what you were thinking?
```python
if args.sim:
        drone_version = c.Drones.LEONARDO_SIM
    else:
        drone_version = c.Drones.LEONARDO
controller = DroneController(drone_version, is_simulation=args.sim)
```

**Comment by ClayMav** [comment]
I was thinking that you should use the is_simulation flag only in the init, and if the DroneController gets is_simulation=True, then the DroneController should do the switching, so similar to what you did, but all within the controller init

**Review by ClayMav — CHANGES_REQUESTED** [review]
As above

**Review by ClayMav — DISMISSED** [review]
Oh wait, that was a fast commit

**Review by christopher-o-toole — DISMISSED** [review]

**Review by christopher-o-toole — DISMISSED** [review]

**Review by ClayMav — DISMISSED** [review]

**Review by ClayMav — APPROVED** [review]

**Review by christopher-o-toole — APPROVED** [review]

**Pull request merged**